packages/dds/sequence/src/intervalCollection.ts

```diff
@@ -1036,11 +1041,113 @@ class EndpointIndex<TInterval extends ISerializableInterval> implements Interval
 	}
 }
 
+class EndInRangeIndex<TInterval extends ISerializableInterval> implements IntervalIndex<TInterval> {
+	private readonly intervalTree;
+
+	constructor(
+		private readonly client: Client,
+		private readonly helpers: IIntervalHelpers<TInterval>,
+	) {
+		// eslint-disable-next-line @typescript-eslint/unbound-method
+		this.intervalTree = new RedBlackTree<TInterval, TInterval>(helpers.compareEnds);
+	}
+
+	public add(interval: TInterval): void {
+		this.intervalTree.put(interval, interval);
+	}
+
+	public remove(interval: TInterval): void {
+		this.intervalTree.remove(interval);
+	}
+
+	/**
+	 * @returns an array of all intervals contained in this collection that locate within the range
+	 * `[startPosition, startPosition + range)`. [5, 10)
+	 */
+	public findIntervalsWithEndInRange(endPosition: number, range: number) {
```

Comment:
before you write too many unit tests, I'd adjust this API how we suggested (it'll be a little annoying to update them all if you do it afterward). The two parameters should be `start` and `end`.